### PR TITLE
Nginx

### PR DESCRIPTION
**Pull request by WCRest**

### CODE REVIEW

File: nginx/hellogov.conf [hunk 1]
```diff
@@ -0,0 +1,19 @@
+server {
+  listen 80;
+  listen [::]:80;
+  server_name hellogov.org;
```
Comment (tjgupta) [inline nginx/hellogov.conf]: Should this now be heyagov.org?

I wonder if (without having chef/ansible/puppet/etc) we need to create a heyagov.conf and a hellogov.conf?

File: update_application.sh [hunk 2]
```diff
@@ -2,23 +2,32 @@
...
+git pull origin master
+
+# load the vars again in case any changed during git pull
+source $REMOTE_SCRIPT_PATH/deploy/load_variables.sh
 
 set -x
 whoami
-##TODO: stop PM2 process -- how do we know the process ID? 
-pm2 stop $REMOTE_SCRIPT_PATH/ecosystem.config.json
+service nginx stop
```
Comment (tjgupta) [inline update_application.sh]: We shouldn't need to completely stop and restart nginx for an application update. Or was there something you ran into that it needed this?

File: update_application.sh [hunk 3]
```diff
@@ -30,5 +39,8 @@ fi
 cp $REMOTE_SCRIPT_PATH/secrets.js ./
 npm install #--production
 #npm prune --production
+npm run build
+npm rebuild node-sass
 
-pm2 start $REMOTE_SCRIPT_PATH/ecosystem.config.json
+pm2 start $REMOTE_SCRIPT_PATH/ecosystem.config.json --env production
+service ngnix start
```
Comment (tjgupta) [inline update_application.sh]: Theoretically shouldn't need.

File: provision_app_server.sh [hunk 4]
```diff
@@ -1,26 +1,41 @@
...
 sudo apt-get install rubygems -y
 sudo apt-get install rubygems-integration -y
 
-sudo iptables -t nat -A PREROUTING -p tcp --dport 80 -j REDIRECT --to-port 8080
+# SSL
+wget https://dl.eff.org/certbot-auto
+chmod a+x certbot-auto
 
-sudo useradd -m hellogov
+# FIXME: don't hard code domain name
```
Comment (tjgupta) [inline provision_app_server.sh]: Tiny thing, but you can remove this comment if you want. I forgot to when I changed to using $DOMAIN_NAME.

File: nginx/hellogov.conf (hunk 1 above)
Comment (WCRest) [reply nginx/hellogov.conf]: I've been putting a placeholder (###DOMAIN_NAME##) in this file and doing a find-replace on it with a domain name the box will serve from. I'll point it out in my next commit. Its not the most robust method, but it lets us deploy to multiple domain names, which is something I've been doing often.

File: update_application.sh (hunk 2 above)
Comment (WCRest) [reply update_application.sh]: You're right. Taking it out.

File: update_application.sh (hunk 3 above)
Comment (WCRest) [reply update_application.sh]: Are you talking about restarting nginx or the production flag to pm2? I'm removing both: 1) you're right, its not necessary to restart nginx and 2) we're specifying NODE_ENV=production elsewhere and for some reason, passing the --env production variable here breaks deployment. Setting the --env production variable adds NODE_ENV=production to the call pm2 makes to start the application.

File: provision_app_server.sh (hunk 4 above)
Comment (WCRest) [reply provision_app_server.sh]: got it.

File: ecosystem.config.json [hunk 5]
```diff
@@ -1,6 +1,12 @@
 {
   "apps" : [{
     "name"        : "hellogov",
-    "script"      : "/var/www/hellogov/app/index.js"
+    "script"      : "/var/www/hellogov/app/index.js",
+    env: {
+      "NODE_ENV": "development",
+    },
+    env_production : {
```
Comment (WCRest) [inline ecosystem.config.json:8]: that --env production parameter in the pm2 call enables these env variables

File: nginx/hellogov.conf [hunk 6]
```diff
@@ -1,15 +1,15 @@
 server {
   listen 80;
   listen [::]:80;
-  server_name hellogov.org;
-  return 301 https://$server_name$request_uri;
+  server_name ###DOMAIN_NAME###;
```
Comment (WCRest) [inline nginx/hellogov.conf]: We find and replace this with `sed` in the provision script.

File: nginx/hellogov.conf [hunk 7]
```diff
@@ -1,15 +1,15 @@
...
+  return 301 https://$server_name$request_uri; ##how do we account for multiple subdomains like www.
 }
 
 server {
   listen 443 ssl;
   listen [::]:443 ssl;
-  ssl_certificate /etc/letsencrypt/live/heyagov.org/fullchain.pem;
-  ssl_certificate_key /etc/letsencrypt/live/heyagov.org/privkey.pem;
+  ssl_certificate /etc/letsencrypt/live/###DOMAIN_NAME###/fullchain.pem;  ## this actually exists at ..../$DOMAIN_NAME/fullchain.pem
```
Comment (WCRest) [inline nginx/hellogov.conf]: certbot writes the key to a directory with the name of the domain we pass it.

File: provision_app_server.sh [hunk 8]
```diff
@@ -2,40 +2,39 @@ source /tmp/load_variables.sh
...
 sudo -u hellogov mkdir -p $ARCHIVE_DIR
 
-git clone $GIT_DEPLOY_URL $REMOTE_SCRIPT_PATH/deploy
+sudo -u hellogov git clone $GIT_DEPLOY_URL $REMOTE_SCRIPT_PATH/deploy
 sudo chown hellogov:hellogov $REMOTE_SCRIPT_PATH/deploy
-sudo chmod u+x *.sh
+sudo chmod o+w $REMOTE_SCRIPT_PATH/deploy
+find $REMOTE_SCRIPT_PATH/deploy/*.sh -type f -exec sudo chmod u+x  {} \; 
+
+sed -i -e 's/###DOMAIN_NAME###/'"$DOMAIN_NAME"'/g' $REMOTE_SCRIPT_PATH/deploy/nginx/hellogov.conf
```
Comment (WCRest) [inline provision_app_server.sh]: Note the find and replace.

File: update_application.sh (hunk 3 above)
Comment (tjgupta) [reply update_application.sh]: I was just talking about the nginx stop/start.

**Comment by timgupta** [comment]
Looks great, I'm just going to rebase one more time to see if I can get rid of my work github commit, then I'll merge.

**Pull request merged**